internal/bsr/bsr.go

```diff
@@ -237,6 +238,140 @@ func OpenSession(ctx context.Context, sessionRecordingId string, f storage.FS, k
 	return session, nil
 }
 
+// Validation provides the results from validating a bsr.
+type Validation struct {
+	SessionRecordingId         string
+	Valid                      bool
+	SessionRecordingValidation *ContainerValidation
+}
+
+// ContainerValidation contains the results from validating a container in a bsr.
+type ContainerValidation struct {
+	Name                    string
+	ContainerType           ContainerType
+	Error                   error
+	FileChecksumValidations ContainerChecksumValidation
+	SubContainers           []*ContainerValidation
+}
+
+// Validate retrieves a BSR from storage using the sessionRecordingId and validates the BSR.
+// All files and sub container files will be verified by comparing it against checksums for
+// each file in SHA256SUM file.
+//
+// Validation will continue even if there's an error encountered during validation.
+// The validation error will be added to the ContainerValidation struct "Error" field for that container.
+func Validate(ctx context.Context, sessionRecordingId string, f storage.FS, keyUnwrapFn kms.KeyUnwrapCallbackFunc) (*Validation, error) {
+	const op = "bsr.Validate"
+
+	switch {
+	case sessionRecordingId == "":
+		return nil, fmt.Errorf("%s: missing session recording id: %w", op, ErrInvalidParameter)
+	case f == nil:
+		return nil, fmt.Errorf("%s: missing storage: %w", op, ErrInvalidParameter)
+	case keyUnwrapFn == nil:
+		return nil, fmt.Errorf("%s: missing key unwrap function: %w", op, ErrInvalidParameter)
+	}
+
+	session, err := OpenSession(ctx, sessionRecordingId, f, keyUnwrapFn)
+	if err != nil {
+		return nil, fmt.Errorf("%s: failed to retrieve session for %s: %w", op, sessionRecordingId, err)
+	}
```

Comment:
We should close the session at the end of this function, ie:

```suggestion
	session, err := OpenSession(ctx, sessionRecordingId, f, keyUnwrapFn)
	if err != nil {
		return nil, fmt.Errorf("%s: failed to retrieve session for %s: %w", op, sessionRecordingId, err)
	}
	defer session.Close(ctx)
```

Reply:
When we open a session, [required fields needed](https://github.com/hashicorp/boundary/blob/ef3954c561ec693ce476550fab3e79e8c3068cc3/internal/bsr/container.go#L510) to close the session are `nil` so closing a session/connection/channel results in panics.
Files like `container.journal` or  `container.sigs` are all `nil` when we open a container. Only `container.container` is not nil.

Instead of calling defer session.Close(ctx), perhaps we could call `session.container.container.Close()` to directly close the opened container?

Also, that is one of the things I realized in some parts of the code. We don't close opened sessions, connections or channels.